### PR TITLE
[WFLY-19324] Anonymous 'captor' service does not reliably provide Ope…

### PR DESCRIPTION
**Pull request by bstansberry**
…nTelemetryConfig; use the DeploymentUnit AttachmentKey mechanism

https://issues.redhat.com/browse/WFLY-19324

Follow up on #17878 to correct a wiring flaw I misunderstood when I reviewed it.

### CODE REVIEW

File: microprofile/telemetry-smallrye/extension/src/main/java/org/wildfly/extension/microprofile/telemetry/MicroProfileTelemetrySubsystemAdd.java [hunk 1]
```diff
@@ -36,12 +30,6 @@ protected void performBoottime(OperationContext context, ModelNode operation, Mo
 
         super.performBoottime(context, operation, model);
 
-        ServiceInstaller.builder(ServiceDependency.on(WildFlyOpenTelemetryConfig.SERVICE_DESCRIPTOR))
-                .withCaptor(openTelemetryConfig::set)
-                .build()
-                .install(context);
```
Comment (pferraro) [inline microprofile/telemetry-smallrye/extension/src/main/java/org/wildfly/extension/microprofile/telemetry/MicroProfileTelemetrySubsystemAdd.java:42]: This installs a "passive" service that captures the WildFlyOpenTelemetryConfig as soon as the service the provides it is started.  While this correctly makes an existing WildFlyOpenTelemetryConfig available to the MicroProfileTelemetryDeploymentProcessor, it is do not _require_ that one is available.  If the WildFlyOpenTelemetryConfig is meant to be a requirement, then this mechanism is insufficient for this use case.  I assume that is the case here?

File: microprofile/telemetry-smallrye/extension/src/main/java/org/wildfly/extension/microprofile/telemetry/MicroProfileTelemetryDependencyProcessor.java [hunk 2]
```diff
@@ -17,15 +17,17 @@
...
     public void deploy(DeploymentPhaseContext phaseContext) {
         addDependencies(phaseContext.getDeploymentUnit());
 
-        // Ensure the OpenTelemetryConfig is available before the next phase DeploymentUnitPhaseService starts
-        phaseContext.requires(ServiceDependency.on(ServiceName.parse(WildFlyOpenTelemetryConfig.SERVICE_DESCRIPTOR.getName())));
+        // Ensure the OpenTelemetryConfig is available to the Phase.POST_MODULE MicroProfileTelemetryDeploymentProcessor
+        // TODO WFCORE-6491 the kernel should support an API such that an OSH can record this requirement without
+        // needing to involve a DUP like this one that is separate from the one that consumes the dependency.
+        phaseContext.addDeploymentDependency(ServiceName.parse(WildFlyOpenTelemetryConfig.SERVICE_DESCRIPTOR.getName()),
+                MicroProfileTelemetryDeploymentProcessor.CONFIG_ATTACHMENT_KEY);
```
Comment (pferraro) [inline microprofile/telemetry-smallrye/extension/src/main/java/org/wildfly/extension/microprofile/telemetry/MicroProfileTelemetryDependencyProcessor.java:30]: Use the cached ServiceName.
```java
phaseContext.addDeploymentDependency(ServiceNameFactory.resolve(WildFlyOpenTelemetryConfig.SERVICE_DESCRIPTOR), MicroProfileTelemetryDeploymentProcessor.CONFIG_ATTACHMENT_KEY);

**Review by pferraro — APPROVED** [review]
Given the existing mechanisms, this is the correct approach (of the two).

File: microprofile/telemetry-smallrye/extension/src/main/java/org/wildfly/extension/microprofile/telemetry/MicroProfileTelemetrySubsystemAdd.java (hunk 1 above)
Comment (bstansberry) [reply microprofile/telemetry-smallrye/extension/src/main/java/org/wildfly/extension/microprofile/telemetry/MicroProfileTelemetrySubsystemAdd.java:42]: Yes, the WildFlyOpenTelemetryConfig is required. There's also no guarantee that the passive service here is started before the DeploymentUnitPhaseService that runs MicroProfileTelemetryDeploymentProcessor is started.

File: microprofile/telemetry-smallrye/extension/src/main/java/org/wildfly/extension/microprofile/telemetry/MicroProfileTelemetryDependencyProcessor.java (hunk 2 above)
Comment (bstansberry) [reply microprofile/telemetry-smallrye/extension/src/main/java/org/wildfly/extension/microprofile/telemetry/MicroProfileTelemetryDependencyProcessor.java]: Thanks for the reminder! Fixed.

**Comment by bstansberry** [comment]
Thanks @pferraro

**Pull request merged**